internal/querynodev2/segments/retrieve.go

```diff
@@ -18,31 +18,71 @@ package segments
 
 import (
 	"context"
+	"fmt"
+	"sync"
 
+	"github.com/milvus-io/milvus-proto/go-api/commonpb"
 	"github.com/milvus-io/milvus/internal/proto/segcorepb"
 	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/pkg/metrics"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+	"github.com/milvus-io/milvus/pkg/util/timerecord"
 	. "github.com/milvus-io/milvus/pkg/util/typeutil"
 )
 
 // retrieveOnSegments performs retrieve on listed segments
 // all segment ids are validated before calling this function
 func retrieveOnSegments(ctx context.Context, manager *Manager, segType SegmentType, plan *RetrievePlan, segIDs []UniqueID, vcm storage.ChunkManager) ([]*segcorepb.RetrieveResults, error) {
-	var retrieveResults []*segcorepb.RetrieveResults
+	var (
+		resultCh = make(chan *segcorepb.RetrieveResults, len(segIDs))
+		errs     = make([]error, len(segIDs))
+		wg       sync.WaitGroup
+	)
 
-	for _, segID := range segIDs {
-		segment, _ := manager.Segment.Get(segID).(*LocalSegment)
-		if segment == nil {
-			continue
-		}
-		result, err := segment.Retrieve(ctx, plan)
+	searchLabel := metrics.SealedSegmentLabel
+	if segType == commonpb.SegmentState_Growing {
+		searchLabel = metrics.GrowingSegmentLabel
+	}
+
+	for i, segID := range segIDs {
+		wg.Add(1)
+		go func(segID int64, i int) {
```

Comment:
Recommend to use `errgroup`, which will stop as saw the first error, also make the code simpler. But LGTM

Reply:
Ok, I'll modify it as well as the errs in search.go in the next pr